lib/exceptions.ts

```diff
@@ -1,55 +1,107 @@
+type Message = string;
+
+interface Status {
+  status: number;
+  statusText: string;
+}
+
+interface ErrorDetails {
+  signature?: string;
+  raw?: any;
+}
+
+interface FetchErrorDetails extends Status {
+  headers: Headers;
+  body: string;
+}
+
+// Deprecated
+interface AxiosErrorDetails extends Partial<Status> {
+  originalError: Error;
+  code?: string;
+}
+
 export class SignatureValidationFailed extends Error {
-  constructor(
-    message: string,
-    public signature?: string,
-  ) {
+  public signature?: string;
+
+  constructor(message: Message, { signature }: ErrorDetails = {}) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { signature });
   }
 }
 
 export class JSONParseError extends Error {
+  public raw: any;
+
+  constructor(message: Message, { raw }: ErrorDetails = {}) {
+    super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { raw });
+  }
+}
+
+export class HTTPFetchError extends Error {
+  public status: number;
+
+  public statusText: string;
+
+  public headers: Headers;
+
+  public body: string;
+
   constructor(
-    message: string,
-    public raw: any,
+    message: Message,
+    { status, statusText, headers, body }: FetchErrorDetails,
   ) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { status, statusText, headers, body });
   }
 }
 
+/* Deprecated */
 export class RequestError extends Error {
-  constructor(
-    message: string,
-    public code: string,
-    private originalError: Error,
-  ) {
+  public code: string;
+
+  private originalError: Error;
+
+  constructor(message: Message, { code, originalError }: AxiosErrorDetails) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { code, originalError });
   }
 }
 
 export class ReadError extends Error {
-  constructor(private originalError: Error) {
-    super(originalError.message);
+  public originalError: Error;
+
+  constructor(message: Message, { originalError }: AxiosErrorDetails) {
+    super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { originalError });
   }
 }
 
 export class HTTPError extends Error {
+  public status: number;
+
+  public statusText: string;
```

Comment:
Please keep `statusMessage` (same reason as `statusCode`) in `HTTPError`

Reply:
I'll double check the changes and return the changed field name of the Deprecated Class! :)

Reply:
Sorry, I didn't understand the status field naming of `HTTPFetchError`.
What field name should I use? :)

```shell
- 1. status
- 2. statusCode
```


+ The naming of the status field of the depreciated `HTTPError` has been restored to the 'statusCode' as your feedback thx! :)

Reply:
The Response class uses `status` to handle the HTTP status code. I also think `statusCode` is clear, and `httpStatusCode` would be fine too. I have no particular preference. Considering that it is `status: Number`, it is not too ambiguous that this represents the HTTP status code.

Let's do this: Using `status` is fine since it's consistent with the Response class. However, if @pengooseDev, as one of an SDK user, prefers it, then `statusCode` is equally good. I will approve either.